app/src/main/java/dev/leonlatsch/photok/model/io/ThumbnailManager.kt

```diff
@@ -0,0 +1,132 @@
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package dev.leonlatsch.photok.model.io
+
+import android.content.Context
+import android.graphics.Bitmap
+import androidx.core.graphics.drawable.toBitmap
+import coil.ImageLoader
+import coil.decode.VideoFrameDecoder
+import coil.request.ImageRequest
+import coil.size.Scale
+import dev.leonlatsch.photok.model.database.entity.Photo
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+
+
+/** Maximum size of the thumbnail in pixels */
+private const val THUMBNAIL_SIZE = 128
+
+/**
+ * Manages the creation and loading of thumbnails for photos and videos.
+ *
+ * @param context the application context
+ * @param encryptedStorageManager the [EncryptedStorageManager] to access the encrypted files
+ *
+ * @since 1.7.2
+ * @author Starry Shivam
+ */
+class ThumbnailManager(
+    private val context: Context,
+    private val encryptedStorageManager: EncryptedStorageManager
+) {
+    /**
+     * Loads the full size thumbnail stored for this photo as a [ByteArray]
+     *
+     * @param photo the photo to load the thumbnail for
+     * @return the thumbnail as a [ByteArray]
+     */
+    fun loadThumbnail(photo: Photo): ByteArray? {
+        encryptedStorageManager.internalOpenEncryptedFileInput(photo.internalThumbnailFileName)
+            ?.use { return it.readBytes() }
+
+        return null
+    }
+
+    /**
+     * Load the full size preview for a stored photo as a [ByteArray]
+     *
+     * @param photo the photo to load the preview for
+     * @return the preview as a [ByteArray]
+     */
+    fun loadVideoPreview(photo: Photo): ByteArray? {
+        encryptedStorageManager.internalOpenEncryptedFileInput(photo.internalVideoPreviewFileName)
+            ?.use { return it.readBytes() }
+
+        return null
+    }
+
+    /**
+     * Creates a thumbnail for the given photo. If the photo is a video,
+     * it also creates a video preview.
+     *
+     * @param photo The photo object for which the thumbnail is to be created.
+     * @param obj The data for the image request. This could be a URL, file, or any other supported data type.
+     * @return A [Result] indicating the success or failure of the thumbnail creation.
+     */
+    suspend fun createThumbnail(
+        photo: Photo, obj: Any?
+    ): Result<Unit> = withContext(Dispatchers.IO) {
+        val deferredResult = CompletableDeferred<Result<Unit>>()
```

Comment:
I like this approach but I would much prefer to use `suspendCorutine` and resume when a result is available. Its used like this in other parts of the app.

```kotlin
): Result<Unit> = suspendCoroutine { continuation ->
...
continuation.resume(Result.success(Unit))
...
continuation.resume(Result.error(e))
```

Reply:
> I like this approach but I would much prefer to use `suspendCorutine` and resume when a result is available. Its used like this in other parts of the app.
> 
> ```kotlin
> ): Result<Unit> = suspendCoroutine { continuation ->
> ...
> continuation.resume(Result.success(Unit))
> ...
> continuation.resume(Result.error(e))
> ```

The problem with this approach is that if you look at the signature of `suspendCoroutine`:

```kotlin
public suspend inline fun <T> suspendCoroutine(
    crossinline block: (Continuation<T>) -> Unit
): 
```

you see that it doesn't accept a _suspendable_ block, since you're supposed to manage suspension yourself via `continuation`, but the `imageLoader.execute(request)` from Coil is a suspend function which cannot be called inside the `suspendCoroutine` block.

Reply:
I see. Didn't think about that 👍